### PR TITLE
feature/store-access-to-aicsimage-and-raw-meta-in-layer-meta

### PR DESCRIPTION
**Pull request by JacksonMaxfield**
**Pull request recommendations:**
- [x] Name your pull request _your-development-type/short-description_. Ex: _feature/read-tiff-files_
- [x] Link to any relevant issue in the PR description. Ex: _Resolves [gh-12], adds tiff file format support_

https://forum.image.sc/t/napari-file-open-and-metadata-access/63497/12

- [x] Provide context of changes.

* Adds `aicsimage` key to the metadata dict stored on each layer allowing users to access the loaded / AICSImage object directly in napari console
* Adds both "raw" and optionally "ome" metadata in the metadata dict -- @tlambert03 currently this change means that users reading anything besides OME-TIFF and CZIs will not have a metadata viewer. Any suggestions on how to create a new "OME" metadata render tree with a different name for the element tree meta?

- [x] Provide relevant tests for your feature or bug fix.
- [x] Provide or update documentation for any feature added by your pull request.

Thanks for contributing!

Can't tag as reviewers but would love opinions from @tlambert03, @psobolewskiPhD, and @pr4deepr

### CODE REVIEW

File: napari_aicsimageio/core.py [hunk 1]
```diff
@@ -93,8 +93,13 @@ def _get_meta(data: xr.DataArray, img: AICSImage) -> Dict[str, Any]:
         meta["scale"] = tuple(scale)
 
     # Apply all other metadata
-    meta["metadata"] = {"ome_types": img.metadata}
+    img_meta = {"aicsimage": img, "raw": img.metadata}
+    try:
+        img_meta["ome_types"] = img.ome_metadata
```
Comment (tlambert03) [inline napari_aicsimageio/core.py:98]: remind me what exception might we be catching here again?

**Review by tlambert03 — APPROVED** [review]
nice!  will be very handy!
In addition to the screenshot in the readme, might be nice to add a bulleted list of specific keys and their meanings `aicsimage`, `ome_types` and `metadata` (or is it `raw`?  screenshot and code seem to conflict)

File: napari_aicsimageio/core.py (hunk 1 above)
Comment (JacksonMaxfield) [reply napari_aicsimageio/core.py:98]: It _should_ just be `NotImplementedError` but in the case the metadata conversion goes wrong, I think each reader can have it's own error handling during conversion.

**Comment by JacksonMaxfield** [comment]
> In addition to the screenshot in the readme, might be nice to add a bulleted list of specific keys and their meanings aicsimage, ome_types and metadata (or is it raw? screenshot and code seem to conflict)

Agree. I will document!

It's the `layer["metadata"]` that has a Dict of extra stuff

so its `layer["metadata"]["raw"]` for the "raw metadata" and `layer["metadata"]["ome_types"]` for the ome object

**Comment by pr4deepr** [comment]
Thanks for making this @JacksonMaxfield 
The `aicsimage` object access works well for me!

> * Adds both "raw" and optionally "ome" metadata in the metadata dict -- @tlambert03 currently this change means that users reading anything besides OME-TIFF and CZIs will not have a metadata viewer. Any suggestions on how to create a new "OME" metadata render tree with a different name for the element tree meta?

I think a different name for the element tree meta, such as `xml` or as you use now, `raw` should work. 
If you get an `xml` for the metadata, can you convert it into an ome tree using [ome_types](https://github.com/tlambert03/ome-types) library written by @tlambert03 ?

**Comment by pr4deepr** [comment]
Oh yes, when I read a czi file I get this:

image

**Comment by JacksonMaxfield** [comment]
> If you get an xml for the metadata, can you convert it into an ome tree using [ome_types](https://github.com/tlambert03/ome-types) library written by @tlambert03 ?

This how it works in the current version. OME Types will render the ElementTree.... I thoughts that why you were confused?

This change will make it only render the tree if it has a metadata converter to OME OR it is an OME file. BUT now the metadata is accessible in the console -- unless we come up with a method to spawn two tree renderers.

If you are asking if we can convert the raw metadata into OME thats the whole deal of like "needing a converter" -- CZI has a converter which is XSLT, bioformatsReader has a converter which is bioformats, but not all readers have converters.

And yes @psobolewskiPhD that is exactly correct.

**Comment by psobolewskiPhD** [comment]
Oh snap! I think I *get it* now!
```
viewer.layers[0].metadata["aicsimage"].scenes
Out[4]: ('A3 Region2_Merged', 'A3 Region1_Merged')
viewer.layers[0].metadata["aicsimage"].dims.order
Out[5]: 'TCZYX'
```
This is pretty sick!
Very cool—the layer data is actually squeezed of those extra dims, so having this info here is quite cool.
:heart:

**Comment by JacksonMaxfield** [comment]
Seems like I definitely need more documentation in the README as to "how this works"...

Taking suggestions but how does this sound:
```
To access the AICSImage object used to load the image data and metadata, you can do the following in the console:

img = viewer.layers[0].metadata["aicsimage"]

This is a normal AICSImage object so all the normal operations apply:

img.dims.order  # TCZYX
img.channel_names  # ["Bright", "Struct", "Nuc", "Memb"]
img.get_image_data("ZYX")  # np.ndarray

The metadata dict also stores shorthands to two variants of the metadata:

viewer.layers[0].metadata["raw"]

Stores the metadata in whichever format is used by the underlying file format reader, i.e. for CZI it returns an xml.ElementTree, for OME-TIFF it returns an OME object from ome-types.

Lastly, if the underlying file format reader has an OME metadata converter, you may additionally see a key in the metadata dict called "ome_types". For example, because the aicsimageio CZIReader and BioformatsReader both support converting metadata to OME metadata, you will see an "ome_types" key.

viewer.layers[0].metadata["ome_types"]  # OME object from ome-types
```

?

---

Again, I want to emphasize, that this PR currently removes the ability to render the ElementTree in the napari viewer. ome-types napari plugin specifically looks at the `"ome_types"` metadata key for metadata pickup, it just so happened that it allowed rendering of an ElementTree in addition to an OME object, and so we were overloading the functionality there. If this PR gets merged as is, LIF files for example wont have a metadata tree render because the ElementTree metadata has been moved from "ome-types" to "raw".

**Comment by pr4deepr** [comment]
> This how it works in the current version. OME Types will render the ElementTree.... I thoughts that why you were confused?

Thanks for clarifying. Yea, I wasn't clear on how the current implementation works, only that it gets an ElementTree.

**Comment by pr4deepr** [comment]
Small suggestion for the documentation, perhaps preface with all the changes (if you haven't already):

The metadata `dict `now currently stores a reference to the AICSImage object used to load the image data and image metadata.

**Comment by JacksonMaxfield** [comment]
> Actually, I just realised. I am not getting an `ome_types` key for my `czi` file. I only get the `raw` key. I think its got to do with the czi file being from the Zeiss lattice microscope which has different metadata compared to traditional `czi` files.

Yep. and for that file you also don't get the render anymore... Is that okay?
If you want to fix that btw you can make a PR to: https://github.com/AllenCellModeling/czi-to-ome-xslt

**Comment by pr4deepr** [comment]
Thanks @JacksonMaxfield . Its not a big issue now. 
I'll make a PR at [czi-to-ome-xslt](https://github.com/AllenCellModeling/czi-to-ome-xslt) as you mentioned. 
Appreciate the prompt responses!

**Comment by JacksonMaxfield** [comment]
Alright, README is updated. Hope the new language clears up any confusion.

Before I merge this, I want to double check: Everyone is okay with me basically removing the metadata tree renderer for any file that isn't itself an OME file or doesn't have an OME metadata transformer? LIF metadata would only be accessible via terminal, ND2 metadata would only be accessible via terminal, etc.

**Comment by tlambert03** [comment]
> Before I merge this, I want to double check: Everyone is okay with me basically removing the metadata tree renderer for any file that isn't itself an OME file or doesn't have an OME metadata transformer? LIF metadata would only be accessible via terminal, ND2 metadata would only be accessible via terminal, etc.

I'm fine with this, and also, if it's possible, I would be happy to update the ome-types tree renderer to look for these keys if there's a clear way they can be shown?  
was LIF or ND2 working anyway with the OME tree viewer?  is there anything I can do to make them work?

**Comment by JacksonMaxfield** [comment]
> I'm fine with this, and also, if it's possible, I would be happy to update the ome-types tree renderer to look for these keys if there's a clear way they can be shown?
was LIF or ND2 working anyway with the OME tree viewer? is there anything I can do to make them work?

Oh weird... I just installed the latest version `napari-aicsimageio` and I guess I was just completely mistaken I don't see the metadata render on LIF or CZI :woman_shrugging:. Now I am confused but I guess this PR can be merged now without any complaint from me :joy:

**Comment by JacksonMaxfield** [comment]
That said, I am going to update the "raw" metadata dict key to "raw_image_meta" so its a _bit_ more clear what is stored there. If you wanted to render it, cool but no worries from me.

**Comment by JacksonMaxfield** [comment]
This has been released. Install with `napari-aicsimageio>=0.6.0`

Thanks all :heart:

**Comment by pr4deepr** [comment]
Thanks for being really quick with this!

**Pull request merged**